### ISSUE TITLE
5.6 Packet Number Protection is hard to follow

### ISSUE THREAD

**Issue by tatsuhiro-t** [quicwg/base-drafts#1387]
It is hard to understand TLS 5.6 Packet Number Protection.

More specifically, this text block:

```
   In sampling the packet ciphertext, the packet number length is
   assumed to be the smaller of the maximum possible packet number
   encoding (4 octets), or the size of the protected packet minus the
   minimum expansion for the AEAD.  For example, the sampled ciphertext
   for a packet with a short header can be determined by:

   "sample_offset = min(1 + connection_id_length + 4, packet_length -
   aead_expansion) sample =
   packet[sample_offset..sample_offset+sample_length] "
```

It is ambiguous that what packet_length is.  Is it the Packet length field in Long packet header?  But it does not appear in Short header.  Is it the length of (Protected) Payload?  Or the length of QUIC packet, or UDP packet?

The text says "the packet number length is ...", but the example seems to be offset to the byte to sample.

It would be also nice to show the example of Long packet.

**Comment by martinthomson** [comment]
The issue here is that it is impossible to find the end of the packet because the end of the packet is used to determine what octets to sample for packet number protection, and the length of the packet number is unknown until packet number protection is removed.  This was fixed by including the packet number in the length field.  This will require notes in the change log.

**Comment by elenagurevich** [comment]
Hello, i wondering to understand how this issue is cleared in case of short header format ?
Long Header just contzine explicit length field may be used for packet sample, what about short header ?

**Comment by martinthomson** [comment]
A UDP datagram has a length.  The short packet takes up the remainder of the UDP datagram.

**Comment by elenagurevich** [comment]
So, UDP datagram contains one and only one QUIC short packet ? 
In case the payload is not long enough we can sample ICV generated by AEAD ?

**Comment by mikkelfj** [comment]
ICV (integrity check value?)
If a short header does not have a valid AEAD tag at the end, it would be protocol error that likely results in a dropped packet, not something to search for regardles of whether there are long packets before, or not. However QUIC states that packets should be processed individually, so a bad short packet should not stop you from processing what comes before (not sure I agree with that, but that is the spec).

**Comment by elenagurevich** [comment]
No i mean the following:
a single QUIC short packet may be contained within a single UDP datagram --> hence UDP header length may be used to determine protected packet length. 
packet_length = sizesof(ShortHeader) + sizeof(Protected_Payload)+sizeof(Authentication Tag)
but if  
_sample_offset + sample_length > packet_length 
then 
sample_offset = packet_length - sample_length_ 
So it may happen  that sample will take bytes from authentication tag (ICV) depending on actual payload size. 
Once the sample is known we can remove protection from the packet_number.
Once packet_number is decrypted,  its size is known, its value can be decoded, so we can form nonce, locate start of protected payload and decrypt& verify authentication tag of a packet.

**Comment by elenagurevich** [comment]
Can you send me a pseudocode reference please ? 
I am new in QUIC so probably just missed all interesting stuff :)

**Comment by martinthomson** [comment]
There is pseudocode about packet number encodings, nothing else (yet).

The important thing about short packet headers is that they can be in the same packet as another packet, but they are always the last packet in the datagram.

**Comment by mikkelfj** [comment]
you mean "can be in the same datagram as another packet..."

pseudocode
https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#rfc.appendix.A
